src/mlpack/methods/hmm/hmm.hpp

```diff
@@ -290,6 +290,73 @@ class HMM
    */
   double LogLikelihood(const arma::mat& dataSeq) const;
 
+  /**
+   * Compute the log of the scaling factor of the given emission probability
+   * at time t. To calculate the log-likelihood for the whole sequence,
+   * accumulate log scale over the entire sequence
+   *
+   * @param emissionLogProb emission probability at time t.
+   * probability up to time t-1
+   * @param prevForwardLogProb Vector in which forward probabilities for time
+   * t-1 will be saved. Passing prevForwardProb as an empty vector indicates the
+   * start of sequence or time t=0
+   * @param forwardLogProb Vector in which forward probabilities for time t
+   * will be saved.
+   * @return Log scale factor of the given sequence of emission at time t.
+   */
+  double LogScaleEmissionProb(const arma::vec& emissionLogProb,
+                       arma::vec& prevForwardLogProb,
+                       arma::vec& forwardLogProb) const;
+  /**
+   * Compute the log-likelihood of the given emission probability up to time t
+   *
+   * @param emissionLogProb emission probability at time t.
+   * @param logLikelihood Log-likelihood of the given sequence of emission
+   * probability up to time t-1
+   * @param prevForwardLogProb Vector in which forward probabilities for time
+   * t-1 will be saved. Passing prevForwardProb as an empty vector indicates the
+   * start of sequence or time t=0
+   * @param forwardLogProb Vector in which forward probabilities for time t
+   * will be saved.
+   * @return Log-likelihood of the given sequence of emission up to time t.
+   */
+  double LogLikelihoodEmissionProb(const arma::vec& emissionLogProb,
+                       double &logLikelihood,
+                       arma::vec& prevForwardLogProb,
+                       arma::vec& forwardLogProb) const;
+  /**
+   * Compute the log of the scaling factor of the given data at time t.
+   * To calculate the log-likelihood for the whole sequence, accumulate log
+   * scale over the entire sequence
+   *
+   * @param data observation at time t.
+   * @param prevForwardLogProb Vector in which forward probabilities for time
+   * t-1 will be saved. Passing prevForwardProb as an empty vector indicates the
+   * start of sequence or time t=0
+   * @param forwardLogProb Vector in which forward probabilities for time t
+   * will be saved.
+   * @return Log scale factor of the given sequence of data up at time t.
+   */
+  double LogScale(const arma::vec &data,
+                       arma::vec& prevForwardLogProb,
+                       arma::vec& forwardLogProb) const;
+  /**
+   * Compute the log-likelihood of the given data up to time t
```

Comment:
I think it might be worth pointing out that this is incremental, so `up to time t` only makes sense in certain situations.  I might suggest documentation like this:

"Compute the log-likelihood of the given observation in `data`, storing the result in `logLikelihood`.  This is meant for incremental or streaming computation of the log-likelihood of a sequence.  For the first data point, provide an empty `prevForwardLogProb` vector.  For subsequent calls, use `forwardLogProb` from the previous call as `prevForwardLogProb`."

I'm not sure if that's exactly the best way to put it, but we should be pretty clear about how to use this function since it's not entirely trivial.

Reply:
sure